### ISSUE TITLE
Scrapy chokes on HTTP response status lines without a Reason phrase

### ISSUE THREAD

**Issue by tonal** [scrapy/scrapy#345]
Try fetch page:

```
$ scrapy fetch 'http://www.gidroprofmontag.ru/bassein/sbornue_basseynu'
```

output:

```
2013-07-11 09:15:37+0400 [scrapy] INFO: Scrapy 0.17.0-304-g3fe2a32 started (bot: amon)
/home/tonal/amon/amon/amon/downloadermiddleware/blocked.py:6: ScrapyDeprecationWarning: Module `scrapy.stats` is deprecated, use `crawler.stats` attribute instead
  from scrapy.stats import stats
2013-07-11 09:15:37+0400 [amon_ra] INFO: Spider opened
2013-07-11 09:15:37+0400 [amon_ra] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2013-07-11 09:15:37+0400 [amon_ra] ERROR: Error downloading <GET http://www.gidroprofmontag.ru/bassein/sbornue_basseynu>: [<twisted.python.failure.Failure <class 'scrapy.xlib.tx._newclient.ParseError'>>]
2013-07-11 09:15:37+0400 [amon_ra] INFO: Closing spider (finished)
2013-07-11 09:15:37+0400 [amon_ra] INFO: Dumping Scrapy stats:
        {'downloader/exception_count': 1,
         'downloader/exception_type_count/scrapy.xlib.tx._newclient.ResponseFailed': 1,
         'downloader/request_bytes': 256,
         'downloader/request_count': 1,
         'downloader/request_method_count/GET': 1,
         'finish_reason': 'finished',
         'finish_time': datetime.datetime(2013, 7, 11, 5, 15, 37, 512010),
         'log_count/ERROR': 1,
         'log_count/INFO': 4,
         'scheduler/dequeued': 1,
         'scheduler/dequeued/memory': 1,
         'scheduler/enqueued': 1,
         'scheduler/enqueued/memory': 1,
         'start_time': datetime.datetime(2013, 7, 11, 5, 15, 37, 257898)}
2013-07-11 09:15:37+0400 [amon_ra] INFO: Spider closed (finished)
```

**Comment by lbsweek** [comment]
I got this error message when using vpn,  I capture wireshark and find there is no response.

**Comment by dangra** [comment]
@lbsweek what do you mean by "no response"? an empty reply without even a first line?

**Comment by dangra** [comment]
After the failed attempt to fix this issue in #1140, I think the only viable approach is a monkeypatch _similar_ to what @tonal proposes in  https://github.com/scrapy/scrapy/issues/345#issuecomment-41649779

**Comment by tonal** [comment]
On latest scrapy ubuntu package (0.25.0-454-gfa1039f+1429829085) I receive similar errors:
```
$ scrapy fetch http://only.ru/catalog/electro_oven/hiddenheater/
...
2015-06-05 12:39:39.5292+0600 [amon_ra] INFO: Spider opened
2015-06-05 12:39:39.7123+0600 [amon_ra] ERROR: Error downloading <GET http://only.ru/catalog/electro_oven/hiddenheater/>: [<twisted.python.failure.Failure <class 'twisted.web._newclient.ParseError'>>]
2015-06-05 12:39:39.7137+0600 [amon_ra] INFO: Closing spider (finished)
```
My monkeypatch work only for scrapy.xlib.tx._newclient.ParseError bat i receive twisted.web._newclient.ParseError.

How to correct path?

**Comment by dangra** [comment]
> My monkeypatch work only for scrapy.xlib.tx._newclient.ParseError bat i receive twisted.web._newclient.ParseError.
> How to correct path?

Monkey patch both. 

For more info on when Scrapy uses one or the other see https://github.com/scrapy/scrapy/blob/master/scrapy/xlib/tx/__init__.py.

**Comment by sunhaowen** [comment]
I added tonal's monkey patch, but I still receive the same error. 

> 2015-12-07 23:47:19 [scrapy] DEBUG: Retrying GET https://api.octinn.com/partner/baidu3600/strategy_list/1 (failed 1 times): [twisted.python.failure.Failure twisted.web._newclient.ParseError: (u'wrong number of parts', 'HTTP/1.1 200')]

below is my code:

```
import scrapy

def here():
    print('PrimeMusic')

def _monkey_patching_HTTPClientParser_statusReceived():
    """
    monkey patching for scrapy.xlib.tx._newclient.HTTPClientParser.statusReceived
    """
    from scrapy.xlib.tx._newclient import HTTPClientParser, ParseError
    old_sr = HTTPClientParser.statusReceived
    def statusReceived(self, status):
        try:
            return old_sr(self, status)
        except ParseError, e:
            if e.args[0] == 'wrong number of parts':
                return old_sr(self, status + ' OK')
            raise
    statusReceived.__doc__ == old_sr.__doc__
    HTTPClientParser.statusReceived = statusReceived

class GiftSpider(scrapy.Spider):

    name = 'gift'
    allowed_domains = ['octinn.com']
    start_urls = [
        'https://api.octinn.com/partner/baidu3600/strategy_list/1'
    ]

    def __init__(self, *args, **kwargs):
        super(GiftSpider, self).__init__(*args, **kwargs)
        _monkey_patching_HTTPClientParser_statusReceived()

    def parse(self, response):
        """

        :param response:
        :return:
        """
        content = response.body
        print(content)
        pass
```

**Comment by liuwwei3** [comment]
@sunhaowen  Great thx， it works！你是monkey请来的救兵吗？

**Comment by leearic** [comment]
这里尽然有好多好多 CHINESE 


leearic@126.com
 
From: liuwwei3
Date: 2015-12-25 14:59
To: scrapy/scrapy
Subject: Re: [scrapy] Error download page: twisted.python.failure.Failure <class 'scrapy.xlib.tx._newclient.ParseError'> (#345)
@sunhaowen Great thx， it works！你是monkey请来的救兵吗？

**Comment by redapple** [comment]
I just found out about https://twistedmatrix.com/trac/ticket/7673
The twisted team is not ready to fix it, unless someone has a real webserver in the wild that does this.

**Comment by kmike** [comment]
@redapple it could also be a bad proxy, not a bad server

**Comment by redapple** [comment]
True.

**Comment by rolando** [comment]
Here is a live example at this time:

```
$ curl -v "http://www.jindai.com.tw/"
> GET / HTTP/1.1
> Host: www.jindai.com.tw
> User-Agent: Mozilla/5.1 (MSIE; YB/9.5.1 MEGAUPLOAD 1.0)
> Accept: */*
> Referer:
>
< HTTP/1.1 200
< Status: 200
< Connection: close
```

**Comment by rolando** [comment]
It seems that this case is common with, for example, custom nginx modules which only set the response status and no reason.

**Comment by rolando** [comment]
Twisted has a patch ready to fix this issue: https://twistedmatrix.com/trac/ticket/7673#comment:5 🎉

**Comment by heyrict** [comment]
I used the monkey patch but the error just turned from
```python
twisted.python.failure.Failure twisted.web._newclient.ParseError: (u'wrong number of parts', 'HTTP/1.1 301')
```
to
```python
twisted.web._newclient.ResponseFailed: [<twisted.python.failure.Failure builtins.TypeError: can't concat bytes to str>]
```
what's wrong?

**Comment by kmike** [comment]
Fixed in Twisted 17.5.0. 

Example websites from this ticket work for me with Scrapy 1.4.0 and Twisted 17.5.0, so I'm closing it. Thanks everyone!

**Comment by mohanbe** [comment]
Basically Scrapy Ignores 404 Error by Default, It was defined httperror middleware.

So, Add **HTTPERROR_ALLOW_ALL = True** to your settings file.

After this you can access **response.status** through your **parse function**.

**Comment by lycanthropes** [comment]
Hi, everyone, now I also met this problem.  My goal is to download pdf files from websites ( such as "http://114.251.10.201/pdf/month?reportId=462837&isPublic=true") ,but I can not download these pdf files completely with scrapy downloadmiddleware  ( using this method, I found the size of many pdf files downloaded is 1KB ), so I turned to stream method in requests.get function [[https://github.com/scrapy/scrapy/issues/3880](url) . But now when I running it,  scrapy often get choked and says "[urllib3.connectionpool] DEBUG: http://114.251.10.201:80 'GET /pdf/month?reportId=128520&isPublic=true HTTP/1.1'"  200 None . It looks there is no failure but scrapy just chokes for several hours.  Any suggestions?
